Reposition caret after an incorrect command #205

edmundmok opened the pull request:
Fixes #205 

Replace use of events to signal incorrect command errors with IncorrectCommandException

damithc commented on src/main/java/seedu/address/logic/commands/CommandResult.java:
```diff
@@ -5,10 +5,12 @@
     public final String feedbackToUser;
 
-    public CommandResult(String feedbackToUser) {
+    public CommandResult(boolean isValidCommand, String feedbackToUser) {
```
Quick comment: since true is the most common value for this parameter, we can have one constructor without the parameter (implied `true`). i.e.
```java
public CommandResult(String feedbackToUser){
    this(true, feedbackToUser);
}
```

edmundmok replied:
Noted, was thinking of this as well.

edmundmok commented:
After replacing `IncorrectCommandAttemptedEvent` with `isValidCommand` in `CommandResult`, there is no longer any need to restore previous commands or restore caret position manually.

pyokagan commented on src/main/java/seedu/address/logic/commands/CommandResult.java:
```diff
@@ -5,10 +5,17 @@
+        this(true, feedbackToUser);
+    }
+
+    public CommandResult(boolean isValidCommand, String feedbackToUser) {
```
What can I do if I want to attach additional information to a failed `CommandResult`?

E.g. The command failure may have been caused by an exception, so I would like to record the exception in the `CommandResult` (the "cause"). This may be useful for detailed error reporting e.g. with stack traces.

Oh speaking of which, maybe we can have a convenience `CommandResult(Throwable e)` constructor which will grab the error message from the exception.

damithc replied:
Good point @pyokagan. The current implementation is a bit limited in that sense because it assumes any additional information is already encoded in the 'feedbackToUser'. Making it more flexible can be done in this PR or as a separate PR.

damithc replied:
@pyokagan @edmundmok I'm ok with changing the code to use exceptions when the execution fails due to user error (e.g. wrong command) or environment error. Perhaps we do that in a separate issue?

pyokagan replied:
> I'm ok with changing the code to use exceptions when the execution fails due to user error (e.g. wrong command) or environment error.

Great! Let's move forward with this, then. :-D

> Perhaps we do that in a separate issue?

Despite what the PR title says, from https://github.com/se-edu/addressbook-level4/pull/215#issuecomment-267771608 I think @edmundmok 's goal with this PR is to replace the "use events to signal errors" approach with a better one. "Not needing to reposition the caret" is just a side-benefit.

So I think we should just do this directly in this PR. If we do it in another PR, we will just be making @edmundmok take a longer roundabout way, since we will need to undo the changes in `CommandResult` in this PR as "signaling the error in `CommandResult`" and "signaling the error as an Exception" are both different approaches to the same problem.

damithc replied:
Sure. I'm ok with either route.

edmundmok commented:
Updated this PR to focus on replacing events with `IncorrectCommandException`.

damithc commented:
Reviewed 1 of 7 files at r1, 10 of 10 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/main/java/seedu/address/ui/CommandBox.java, line 83 at r2](https://reviewable.io:443/reviews/se-edu/addressbook-level4/215#-K_8-3hMm4vY9I575IDQ:-K_8-3hMm4vY9I575IDR:b-pihwo6) ([raw file](https://github.com/se-edu/addressbook-level4/blob/9a7289aaa7f4f5e1ca756ae69b2487412f4cd27c/src/main/java/seedu/address/ui/CommandBox.java#L83)):*
> ```Java
>         CommandResult mostRecentResult = logic.execute(previousCommandText);
>         handleCorrectCommandAttempted(mostRecentResult);
>     }
> ```

I think inlining this method can make the `handleCommandInputChanged` easier to understand. Reason: that method already mentions `handleIncorrectCommandAttempted(e)` so it makes sense to mention its counterpart `handleCorrectCommandAttempted(mostRecentResult)` in the same vicinity.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/se-edu/addressbook-level4/215)*
<!-- Sent from Reviewable.io -->

edmundmok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/main/java/seedu/address/ui/CommandBox.java, line 83 at r2](https://reviewable.io:443/reviews/se-edu/addressbook-level4/215#-K_8-3hMm4vY9I575IDQ:-K_8A55g9EHn9457VcGz:br8t8kp) ([raw file](https://github.com/se-edu/addressbook-level4/blob/9a7289aaa7f4f5e1ca756ae69b2487412f4cd27c/src/main/java/seedu/address/ui/CommandBox.java#L83)):*
<details><summary><i>Previously, damithc (Damith C. Rajapakse) wrote…</i></summary><blockquote>

I think inlining this method can make the `handleCommandInputChanged` easier to understand. Reason: that method already mentions `handleIncorrectCommandAttempted(e)` so it makes sense to mention its counterpart `handleCorrectCommandAttempted(mostRecentResult)` in the same vicinity.
</blockquote></details>

I see your point. Will update this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/se-edu/addressbook-level4/215)*
<!-- Sent from Reviewable.io -->

damithc commented:
@edmundmok what's the status of this PR?

edmundmok commented:
@damithc Ready for review.

damithc commented on src/main/java/seedu/address/ui/CommandBox.java:
```diff
@@ -65,19 +64,26 @@ public void setPlaceholder(AnchorPane pane) {
     @FXML
     private void handleCommandInputChanged() {
-        //Take a copy of the command text
+        updatePreviousCommandText();
```
Do we still need to keep track of the previous command?

edmundmok replied:
It is definitely unnecessary in the context of the entire class, but will be used in the logging. Can be made into just a local variable instead.

edmundmok replied:
Updated the PR.

damithc commented on src/main/java/seedu/address/ui/CommandBox.java:
```diff
@@ -86,18 +89,11 @@ private void setStyleToIndicateCorrectCommand() {
-        commandTextField.setText(previousCommandText);
+        logger.info("Invalid command: " + incorrectCommandText);
+        raise(new NewResultAvailableEvent(e.getMessage()));
     }
```
I tried inlining some methods. How about this?
```java
    @FXML
    private void handleCommandInputChanged() {
        try {
            CommandResult commandResult = logic.execute(commandTextField.getText());
            setStyleToIndicateCorrectCommand();
            commandTextField.setText("");
            logger.info("Result: " + commandResult.feedbackToUser);
            raise(new NewResultAvailableEvent(commandResult.feedbackToUser));
        } catch (IncorrectCommandException e) {
            setStyleToIndicateIncorrectCommand();
            logger.info("Invalid command: " + commandTextField.getText());
            raise(new NewResultAvailableEvent(e.getMessage()));
        }
    }

    /**
     * Sets the command box style to indicate a correct command.
     */
    private void setStyleToIndicateCorrectCommand() {
        commandTextField.getStyleClass().remove("error");
    }

    /**
     * Sets the command box style to indicate an error
     */
    private void setStyleToIndicateIncorrectCommand() {
        commandTextField.getStyleClass().add("error");
    }
```

edmundmok replied:
Hi Prof, wouldn't it be better to SLAP the code within the try and catch blocks for easier reading?

damithc replied:
Each line is a distinct step of the process, roughly at the same level. So either we can abstract the two blocks into two methods, or each line into a separate method, neither of which seem to help much.

edmundmok replied:
Would this be alright?
```java
    @FXML
    private void handleCommandInputChanged() {
        try {
            CommandResult commandResult = logic.execute(commandTextField.getText());
            handleCorrectCommandAttempted(commandResult);
        } catch (IncorrectCommandException e) {
            handleIncorrectCommandAttempted(e);
        }
    }
```
I feel that the handling parts can be grouped together as one method.

damithc replied:
Possibly. It looks neat and good code usually looks neat. I feel 'handle' is a contrived/forced abstraction though. Try to write header comments for `handleCorrectCommandAttempted` and `handleIncorrectCommandAttempted`. If you have difficulty writing a concise summary sentence that explains what the method does, then that's a sign that it is not a good unit of abstraction.

pyokagan replied:
@damithc 
> It looks neat and good code usually looks neat.

It looks *short*, but it is also very opaque. Readers reading `handleCommandInputChanged()` have to jump outside of this function in order to fully understand what it does.

@edmundmok  Every method (indirection) we extract out incurs a readability cost. Most of the time this is balanced out by other factors (e.g. easier to reason about and improved testability), but I don't think that's the case here (e.g. given the generic name and the methods being private).

Maybe what you want is just a comment? e.g. `// Handle correct command attempted` and `// Handle incorrect command attempted`.

edmundmok replied:
Handling the correct/incorrect command attempted can involve doing many steps, or doing nothing, but here we are allowed to ignore what is actually being done by just asking for it to be handled, so I think it is still a valid abstraction though.

damithc replied:
How much details to include in the 'single level of abstraction' that a method supposed to contain is not very clear cut. It can depend on how think we imagine a slice the abstraction level to be. If you ask what is the highest level logic of this method, the answer could be 'try to execute the command; indicate an error if the command is incorrect', which translates to this.
```java
    void handleCommandInputChanged() {
        try {
            execute(commandTextField.getText());
        } catch (IncorrectCommandException e) {
            indicateIncorrectCommand(e);
        }
    }
```
That follows SLAP, but the slicing is very thin.  The one below contains a thicker slice (Note that it is still evenly spread. Even when we can't agree on thickness, we should ensure the entire slice is the same thickness).

```java
   private void handleCommandInputChanged() {
        try {
            CommandResult commandResult = logic.execute(commandTextField.getText());
            setStyleToIndicateCorrectCommand();
            commandTextField.setText("");
            logger.info("Result: " + commandResult.feedbackToUser);
            raise(new NewResultAvailableEvent(commandResult.feedbackToUser));
        } catch (IncorrectCommandException e) {
            setStyleToIndicateIncorrectCommand();
            logger.info("Invalid command: " + commandTextField.getText());
            raise(new NewResultAvailableEvent(e.getMessage()));
        }
    }
``` 

We can also factor in 'how much details a typical reader of this method is likely to be looking for' to avoid inflicting an unnecessary navigation burden on all readers. 
Still this is a matter of personal preference too. There are those who insist on very thin slices ([this guy](https://robots.thoughtbot.com/sandi-metz-rules-for-developers) advocates less than 5 lines per method, which implies very thin slices). 

As a pragmatic compromise, I sometimes use slightly thicker slices (i.e. about two thin slices of thickness) with interleaving comment lines to explain the intermediate layer, as long as the method is not too long and nesting is not too deep. E.g.

```java
   private void handleCommandInputChanged() {
        try {
            CommandResult commandResult = logic.execute(commandTextField.getText());
            
           // process result of the command
            setStyleToIndicateCorrectCommand();
            commandTextField.setText("");
            logger.info("Result: " + commandResult.feedbackToUser);
            raise(new NewResultAvailableEvent(commandResult.feedbackToUser));

        } catch (IncorrectCommandException e) {

            // handle incorrect command
            setStyleToIndicateIncorrectCommand();
            logger.info("Invalid command: " + commandTextField.getText());
            raise(new NewResultAvailableEvent(e.getMessage()));
        }
    }
``` 

@pyokagan any strong preference of the two choices? I'm OK with both as both are within my tolerance of SLAP levels.

pyokagan replied:
@damithc 
> any strong preference of the two choices?

Hmm? I see three code snippits in your message :-D.

The first one is definitely too opaque as I mentioned above, and I'm strongly against it. The second one is good, but the third one is even better because the `// process result of the command` comment makes it slightly more readable to me. (Although the blank line above the `// handle incorrect command` looks weird.)

edmundmok replied:
Alright, will update to the third one as recommended.

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -105,11 +106,17 @@ private void assertCommandBehavior(String inputCommand, String expectedMessage,
-        CommandResult result = logic.execute(inputCommand);
+        String feedbackToUser;
+        try {
+            //Execute the command
```
Redundant comment?

edmundmok replied:
Yes, will remove!

edmundmok commented on src/main/java/seedu/address/ui/CommandBox.java:
```diff
@@ -86,18 +89,11 @@ private void setStyleToIndicateCorrectCommand() {
         commandTextField.setText("");
```
True, will move it out.

edmundmok commented:
Trying to rebase this commit on the new master, ignore the failed attempt!

edmundmok commented:
Updated the PR. Ready for further review.

pyokagan commented on src/main/java/seedu/address/commons/exceptions/IncorrectCommandException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Signals that an incorrect command was attempted to be executed.
+ */
+public class IncorrectCommandException extends Exception {
```
What's an "incorrect" command?

The name "incorrect" was inherited from the code this is replacing, and it is probably called "incorrect" because it's used for e.g. boundary checks on command inputs which can only be done when the command is actually executed.

However, looking forward, will this exception only be used for checks on the command inputs? For example, we may have totally correct inputs (e.g. a valid file name), but when we attempt to execute the command (e.g. write to a file), we get an `IOException`. This exception should then be properly propagated to the caller via this "command exception" mechanism.

So, I think that this exception should have the more generic name `CommandException`, and the javadoc should be changed to indicate that this exception is thrown whenever *any* non-programmer error occurs when the command is executed.

edmundmok replied:
So, e.g. we have a more generic `CommandException` and then have `IncorrectCommandException` subclass from it?

pyokagan replied:
@edmundmok 
> So, e.g. we have a more generic CommandException and then have IncorrectCommandException subclass from it?

No, just having a `CommandException` will do.

pyokagan commented on src/main/java/seedu/address/commons/exceptions/IncorrectCommandException.java:
```diff
@@ -0,0 +1,10 @@
+package seedu.address.commons.exceptions;
```
Should this really be in the `commons` package, given that this exception only applies to the `logic.commands` package?

Maybe this exception should be in `logic.commands` instead?

edmundmok replied:
Noted.

edmundmok commented:
Updated the PR, also removed the `registerAsEventHandler() `.

pyokagan commented on src/main/java/seedu/address/ui/CommandBox.java:
```diff
@@ -62,42 +59,30 @@ public void setPlaceholder(AnchorPane pane) {
     /**
      * Sets the command box style to indicate a correct command.
      */
     private void setStyleToIndicateCorrectCommand() {
```
@damithc While we are touching this part, do you think we should change the naming to `setStyleToIndicateCommandSuccess()` and `setStyleToIndicateCommandFailure()` ? This correct/incorrect command terminology does not seem appropriate to me, as it implies that the error only occurs because an `IncorrectCommand` is run, when that is not the case (there are lots of other possible factors which could cause a command to fail).

damithc replied:
> do you think we should change the naming to setStyleToIndicateCommandSuccess() and setStyleToIndicateCommandFailure() ?

Good idea @pyokagan

edmundmok replied:
@damithc @pyokagan True. I think valid/invalid command might be more suitable, since this style is only activated upon any invalid command inputs (might be an `IncorrectCommand`, might be a wrong format of a valid command), but not on all command failures. `AddCommand` can still "fail" e.g. due to `DuplicatePersonException` but it will not be captured as a `CommandException` and thus won't be a "command failure", even though the command failed. Well, unless we want to update it to be received as a `CommandException`.

pyokagan replied:
@edmundmok 
> AddCommand can still "fail" e.g. due to DuplicatePersonException but it will not be captured as a CommandException and thus won't be a "command failure", even though the command failed.

:-O Sounds like a bug. It is definitely a command failure if a command which is supposed to be adding persons fails to add a person.

> Well, unless we want to update it to be received as a CommandException.

Yes, it definitely should.

edmundmok replied:
Alright, will update it.

pyokagan commented on src/main/java/seedu/address/logic/commands/IncorrectCommand.java:
```diff
@@ -1,5 +1,6 @@
+import seedu.address.logic.commands.exceptions.CommandException;
 
 /**
  * Represents an incorrect command. Upon execution, produces some feedback to the user.
```
Maybe update this javadoc, e.g. `Upon execution, throws a CommandException with feedback to the user.`

edmundmok replied:
Noted.

pyokagan commented on src/main/java/seedu/address/logic/commands/Command.java:
```diff
@@ -25,8 +24,9 @@ public static String getMessageForPersonListShownSummary(int displaySize) {
      * Executes the command and returns the result message.
      *
      * @return feedback message of the operation result for display
+     * @throws CommandException If the Command is an incorrect command
```
It's not just when the `Command` is an "incorrect command", but whenever an error occurs while executing the command.

edmundmok replied:
Noted.

edmundmok commented on src/main/java/seedu/address/logic/Logic.java:
```diff
@@ -13,7 +14,7 @@
      * @param commandText The command as entered by the user.
      * @return the result of the command execution.
```
Ah, thanks for the reminder!

edmundmok commented on src/main/java/seedu/address/ui/CommandBox.java:
```diff
@@ -62,42 +59,30 @@ public void setPlaceholder(AnchorPane pane) {
+            raise(new NewResultAvailableEvent(commandResult.feedbackToUser));
+
+        } catch (CommandException e) {
+            // handle incorrect command
```
Will update to `// handle command failure`

edmundmok commented:
@pyokagan Updated `AddCommand` to throw `CommandException` when a `DuplicatePersonException` is caught.

pyokagan commented on src/main/java/seedu/address/logic/commands/IncorrectCommand.java:
```diff
@@ -1,8 +1,9 @@
+import seedu.address.logic.commands.exceptions.CommandException;
 
 /**
- * Represents an incorrect command. Upon execution, produces some feedback to the user.
```
The `Represents an incorrect command.` has some value, I think, since it somewhat explains the name `IncorrectCommand`.

pyokagan commented on src/main/java/seedu/address/logic/commands/exceptions/CommandException.java:
```diff
@@ -0,0 +1,10 @@
+package seedu.address.logic.commands.exceptions;
+
+/**
+ * Represents any non-programmer error during the execution of a command.
```
I don't think we need to emphasize `non-programmer` here in the first sentence, although it would indeed be good to recommend that `CommandException` not be abused for programming errors. So something like:
```
Represents an error which occurs during execution of a {@link Command}.
Should not be used for representing programming errors.
```
I dunno. @damithc do you have any advice on how we can phrase this? `non-programmer error` just sounds extremely weird to me.

damithc replied:
The phrasing suggested by @pyokagan is fine. The second sentence can be omitted as checked exceptions are not supposed to indicate programming errors anyway?

edmundmok replied:
Sounds good.

pyokagan commented on src/main/java/seedu/address/ui/CommandBox.java:
```diff
@@ -62,48 +59,37 @@ public void setPlaceholder(AnchorPane pane) {
 
     /**
-     * Sets the command box style to indicate a correct command.
+     * Sets the command box style to indicate a command success.
```
`to indicate a command success` -> `to indicate command success.` or `to indicate a successful command.`

edmundmok replied:
`to indicate a successful command` sounds good.

pyokagan commented on src/main/java/seedu/address/ui/CommandBox.java:
```diff
@@ -62,48 +59,37 @@ public void setPlaceholder(AnchorPane pane) {
 
     /**
-     * Sets the command box style to indicate an error
+     * Sets the command box style to indicate a command failure.
```
`to indicate a command failure` -> `to indicate command failure` or `to indicate a failed command`.

pyokagan commented on src/main/java/seedu/address/logic/Logic.java:
```diff
@@ -12,8 +13,9 @@
      * Executes the command and returns the result.
      * @param commandText The command as entered by the user.
      * @return the result of the command execution.
+     * @throws CommandException If any non-programmer error is encountered during command execution.
```
Maybe we should make this description more generic? After all, callers have no control over whether commands throw `CommandExceptions` in response to programmer errors or not.

Maybe something like:
```
if an error occurs during command execution.
```
Note the use of `an` rather than `any`. We can't guarantee that *any* error which occurs during command execution will actually throw a `CommandException`.

(This goes for the other `@throws` description).

edmundmok replied:
Noted.

edmundmok commented:
@pyokagan Updated as recommended. Thanks for the thorough review.

edmundmok commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -9,7 +9,7 @@
     public void commandBox_commandSucceeds_textCleared() {
         commandBox.runCommand(td.benson.getAddCommand());
-        assertEquals(commandBox.getCommandInput(), "");
+        assertEquals(commandBox.getCommandInput(), td.benson.getAddCommand());
```
@pyokagan Changing the add command to throw a CommandException when trying to add a duplicate requires this to be changed. Will update this further (e.g. the command does not actually succeed so the name is misleading) and let you know.

edmundmok commented:
@pyokagan 
Updated the following: 
1. In `CommandBoxTest`, changed the command in the `commandSucceeds` test to an actual successful one: `select 3`.

2. Also added an `assertCommandBoxInputAndColor` method to test both the command box text and background color. This also means that `getBackgroundColor` was added in `CommandBoxHandle` and `GuiHandle`.

3. In the `commandSucceeds` test of `CommandBoxTest`, I had to use the color code `0x039ed3ff` which looks quite ugly. Not sure how else to go about doing this.

edmundmok commented:
@pyokagan Fair point, I did consider testing for the `error` style class but thought testing the color directly might be more rigorous. Have changed it to test for the `error` instead.

pyokagan commented on src/test/java/guitests/guihandles/GuiHandle.java:
```diff
@@ -66,6 +67,10 @@ protected void setTextField(String textFieldId, String newText) {
     }
 
+    protected ObservableList<String> getStyleClass(String fieldName) {
+        return getNode(fieldName).getStyleClass();
```
Do we need this? Why can't callers just call `getNode(fieldName).getStyleClass()` directly?

edmundmok replied:
For convenience? `getTextField(...)` does this as well. Although, I just realised there might be a typo in `getTextField(String filedName)` -> should be `fieldName`

pyokagan replied:
@edmundmok 
> For convenience?

Well, `getNode(fieldName).getStyleClass()` isn't any longer than `getStyleClass(fieldName)`.

Plus, think of the precedent you are setting here. JavaFX's `Node` class has lots of getter methods, are we going to create a convenience method for every single one of them?

edmundmok replied:
Hmm, that's true.

pyokagan commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -3,20 +3,25 @@
-        assertEquals(commandBox.getCommandInput(), "");
+        commandBox.runCommand("select 3");
+        assertEquals("", commandBox.getCommandInput());
+        assertFalse(commandBox.getStyleClass().contains(ERROR_STYLE));
```
Maybe tighten this test by checking whether the command box *only* adds the error style class?

edmundmok replied:
Good idea.

edmundmok replied:
Think this is for the test case below.

pyokagan commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -3,20 +3,25 @@
-        assertEquals(commandBox.getCommandInput(), "invalid command");
-        //TODO: confirm the text box color turns to red
+        assertEquals("invalid command", commandBox.getCommandInput());
+        assertTrue(commandBox.getStyleClass().contains(ERROR_STYLE));
```
Maybe tighten this test by checking whether the command box removes the error style class *only*?

edmundmok replied:
Good idea.

edmundmok replied:
Wait a minute. This test actually starts out with a command box without the error style and then tests if the error is being added correctly upon an invalid command. But the test scenario that you mention about removing the error style from an initial command box that already has the error style sounds like a good test case that I haven't included yet.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -105,11 +106,16 @@ private void assertCommandBehavior(String inputCommand, String expectedMessage,
+            CommandResult result = logic.execute(inputCommand);
+            feedbackToUser = result.feedbackToUser;
+        } catch (CommandException e) {
+            feedbackToUser = e.getMessage();
```
This is actually slightly icky. By right, we should be checking if the command actually throws a `CommandException`, rather than just relying on the feedback. But yes, I know that almost all of the tests are already coded this way.

Maybe add a `TODO`?

edmundmok commented:
@pyokagan Made some changes according to your comments.

1. Changed `CommandBoxTest` tests to check that only error is added/removed. Code does not look that beautiful however, will need some help improving its quality!

2. Changed `LogicManagerTest`, specifically added a `boolean expectCommandException` parameter into `assertCommandBehavior` so as to check that a command exception is thrown when expected. Not sure if this is the best approach to take. Added convenience methods `assertCommandSuccessBehavior(...)` and `assertCommandFailureBehavior(...)` that in turns calls the new `assertCommandBehavior`. And updated all the previous `assertCommandBehavior` calls to either `assertCommandSuccessBehavior` or `assertCommandFailureBehavior` depending on valid/invalid commands.

pyokagan commented:
@edmundmok 
> Changed LogicManagerTest, specifically added a boolean expectCommandException parameter into assertCommandBehavior so as to check that a command exception is thrown when expected. Not sure if this is the best approach to take.

Knee-jerk reaction: probably not since it reduces the cohesiveness of the `assertCommandBehavior` method.

> Also, I noticed that assertCommandBehavior states that it throws Exception but it does not really throw any exceptions. Should this be removed?

Yeah, it should.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -90,49 +91,82 @@ public void execute_invalid() throws Exception {
      *      - {@code expectedAddressBook} was saved to the storage file. <br>
      */
-    private void assertCommandBehavior(String inputCommand, String expectedMessage,
+    private void assertCommandBehavior(boolean expectCommandException, String inputCommand, String expectedMessage,
```
Oh well, let's just go with this for now.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -90,49 +91,82 @@ public void execute_invalid() throws Exception {
         assertEquals(expectedAddressBook, latestSavedAddressBook);
     }
 
+    private void assertCommandSuccessBehavior(String inputCommand, String expectedMessage,
```
Let's rename this to `assertCommandSuccess`, which is shorter.

Also, try to group overloaded methods together.

edmundmok replied:
Noted.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -90,49 +91,82 @@ public void execute_invalid() throws Exception {
+        assertCommandBehavior(false, inputCommand, expectedMessage, expectedAddressBook, expectedShownList);
+    }
+
+    private void assertCommandFailureBehavior(String inputCommand, String expectedMessage,
```
Let's rename this to `assertCommandFailure`, which is shorter.

edmundmok replied:
Noted.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -90,49 +91,82 @@ public void execute_invalid() throws Exception {
+
+    private void assertCommandFailureBehavior(String inputCommand, String expectedMessage,
+            ReadOnlyAddressBook expectedAddressBook,
+            List<? extends ReadOnlyPerson> expectedShownList) throws Exception {
```
Do we need a variant of `assertCommandFailureBehavior` which allows callers to specify `expectedAddressBook` and `expectedShownList`?

If we think about it, if a command fails, it should always leave the address book and filtered persons list untouched. So, we could implement `assertCommandFailureBehavior` like this:
```java
    private void assertCommandFailureBehavior(String inputCommand, String expectedMessage) throws Exception {
        AddressBook expectedAddressBook = new AddressBook(model.getAddressBook());
        List<ReadOnlyPerson> expectedShownList = new ArrayList<>(model.getFilteredPersonList());
        assertCommandBehavior(true, inputCommand, expectedMessage, expectedAddressBook, expectedShownList);
    }
```
Notice how it calls `assertCommandBehavior` directly -- we should not need to have a `assertCommandFailureBehavior(inputCommand, expectedMessage, expectedAddressBook, expectedShownList)` overload.

edmundmok replied:
That's true. Will update.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -90,49 +91,82 @@ public void execute_invalid() throws Exception {
+                expectedMessage, new AddressBook(), Collections.emptyList());
+    }
+
+    private void assertCommandSuccessBehavior(String inputCommand, String expectedMessage) throws Exception {
```
I think we can do without this overload.

The only times where this is called is in `execute_help()` and `execute_exit()`, and it would probably help if we made those call sites more explicit by letting them provide the `new AddressBook()` and `Collections.emptyList()` directly.

edmundmok replied:
Noted.

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -90,49 +91,82 @@ public void execute_invalid() throws Exception {
      */
-    private void assertCommandBehavior(String inputCommand, String expectedMessage) throws Exception {
-        assertCommandBehavior(inputCommand, expectedMessage, new AddressBook(), Collections.emptyList());
+    private void assertCommandBehavior(boolean expectCommandException, String inputCommand, String expectedMessage)
```
With the introduction of `assertCommandSuccessBehavior()` and `assertCommandFailureBehavior()`, we can probably do without this overload.

edmundmok commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -3,20 +3,79 @@
 public class CommandBoxTest extends AddressBookGuiTest {
 
+    private static final String ERROR_STYLE = "error";
+    private static final String EMPTY_STRING = "";
```
True...

edmundmok commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -3,20 +3,79 @@
-    public void commandBox_commandSucceeds_textCleared() {
-        commandBox.runCommand(td.benson.getAddCommand());
-        assertEquals(commandBox.getCommandInput(), "");
+    public void commandBox_commandSucceeds_textClearedAndNoErrorAdded() {
```
Alright

edmundmok commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -3,20 +3,79 @@
-        commandBox.runCommand(td.benson.getAddCommand());
-        assertEquals(commandBox.getCommandInput(), "");
+    public void commandBox_commandSucceeds_textClearedAndNoErrorAdded() {
+        List<String> styleBeforeCommandExecution = getCurrentStyle();
```
Oh that's a good idea. Will update the other tests to follow this.

edmundmok commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -1,22 +1,49 @@
+
+    @Before
+    public void setCommandBoxStyles() {
+        commandBox.getStyleClass().setAll(DEFAULT_COMMANDBOX_STYLES);
```
@pyokagan Since I am using the default command box styles, I'm not sure if this setup code is necessary actually. However, without this the test will fail if the actual default styles change. This setup sounds pretty weird though. And if I clear the styles in the setup then the GUI will look weird during testing, which is probably not alright.

pyokagan replied:
@edmundmok 
> And if I clear the styles in the setup then the GUI will look weird during testing, which is probably not alright.

Okay, let's go back to doing the previous "constructing the expectedStyleClass list and then compare it" approach.

Something like this on top of your PR?
```diff
diff --git a/src/test/java/guitests/CommandBoxTest.java b/src/test/java/guitests/CommandBoxTest.java
index 7cf16959..ad6b8912 100644
--- a/src/test/java/guitests/CommandBoxTest.java
+++ b/src/test/java/guitests/CommandBoxTest.java
@@ -4,19 +4,22 @@ import org.junit.Before;
 import org.junit.Test;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 
-import java.util.Arrays;
-import java.util.List;
+import java.util.ArrayList;
 
 public class CommandBoxTest extends AddressBookGuiTest {
 
-    private static final List<String> DEFAULT_COMMANDBOX_STYLES = Arrays.asList("text-input", "text-field");
     private static final String COMMAND_THAT_SUCCEEDS = "select 3";
     private static final String COMMAND_THAT_FAILS = "invalid command";
+    private static final String ERROR_STYLE_CLASS = "error";
+
+    private ArrayList<String> expectedStyleClass;
 
     @Before
-    public void setCommandBoxStyles() {
-        commandBox.getStyleClass().setAll(DEFAULT_COMMANDBOX_STYLES);
+    public void setUp() {
+        expectedStyleClass = new ArrayList<>(commandBox.getStyleClass());
+        assertFalse(expectedStyleClass.contains(ERROR_STYLE_CLASS));
     }
 
     @Test
@@ -24,7 +27,7 @@ public class CommandBoxTest extends AddressBookGuiTest {
         commandBox.runCommand(COMMAND_THAT_SUCCEEDS);
 
         assertEquals("", commandBox.getCommandInput());
-        assertEquals(DEFAULT_COMMANDBOX_STYLES, commandBox.getStyleClass());
+        assertEquals(expectedStyleClass, commandBox.getStyleClass());
     }
 
     @Test
@@ -32,7 +35,8 @@ public class CommandBoxTest extends AddressBookGuiTest {
         commandBox.runCommand(COMMAND_THAT_FAILS);
 
         assertEquals(COMMAND_THAT_FAILS, commandBox.getCommandInput());
-        assertEquals(Arrays.asList("text-input", "text-field", "error"), commandBox.getStyleClass());
+        expectedStyleClass.add("error");
+        assertEquals(expectedStyleClass, commandBox.getStyleClass());
     }
 
     @Test
@@ -43,7 +47,7 @@ public class CommandBoxTest extends AddressBookGuiTest {
         commandBox.runCommand(COMMAND_THAT_SUCCEEDS);
 
         assertEquals("", commandBox.getCommandInput());
-        assertEquals(DEFAULT_COMMANDBOX_STYLES, commandBox.getStyleClass());
+        assertEquals(expectedStyleClass, commandBox.getStyleClass());
     }
 
 }
```

edmundmok commented:
@pyokagan Updated the PR according to recommendations for the test! Thanks!

pyokagan commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -1,22 +1,52 @@
-        commandBox.runCommand("invalid command");
-        assertEquals(commandBox.getCommandInput(), "invalid command");
-        //TODO: confirm the text box color turns to red
+    public void commandBox_commandFails_textStaysAndErrorAdded() {
```
Maybe change `textStaysAndErrorAdded` to `textStaysAndErrorStyleClassAdded` to make it more explicit.

edmundmok replied:
Noted.

pyokagan commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -1,22 +1,52 @@
 
 public class CommandBoxTest extends AddressBookGuiTest {
 
+    private ArrayList<String> expectedStyleClass;
```
I think usually the `static` fields come before the instance fields.

edmundmok replied:
Noted.

pyokagan commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -1,22 +1,52 @@
+    }
+
+    @Test
+    public void commandBox_commandSucceedsAfterFailedCommand_textClearedAndErrorRemoved() {
```
Maybe change `textClearedAndErrorRemove` to `textClearedAndErrorStyleClassRemoved` to make it more explicit.

edmundmok replied:
Noted.

edmundmok commented:
@pyokagan PR updated. Thanks!

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -142,28 +168,33 @@ public void execute_clear() throws Exception {
+        assertCommandFailure("add Valid Name p/12345 valid@email.butNoPrefix a/valid, address",
+                expectedMessage);
+        assertCommandFailure("add Valid Name p/12345 e/valid@email.butNoAddressPrefix valid, address",
+                expectedMessage);
```
@pyokagan your thoughts on the below?
In cases like this, I often suggest to adjust indentation manually to make the line wrapping more readable. 

```java
assertCommandFailure("add Valid Name p/12345 e/valid@email.butNoAddressPrefix valid, address",
                                                                               expectedMessage);
```

However, it is a lot of extra work (for the dev to do as well as reviewer to explain :-p) and makes future automated refactoring messier (because the IDE doesn't honor/understand the extra indentation). The alternative is to just accept the default 8 space indentation and live with slightly-harder-to-read code; easier to enforce and less work. 

What do you think?

edmundmok replied:
@damithc Oh, actually some of the lines can be reverted back to not needing line breaks. Forgot to change them back after shortening the `assertCommandFailure` name!

damithc replied:
> Oh, actually some of the lines can be reverted back to not needing line breaks. Forgot to change them back after shortening the assertCommandFailure name!

That's true. Still, it is good to have a consistent policy on his matter.

pyokagan replied:
@damithc 
> In cases like this, I often suggest to adjust indentation manually to make the line wrapping more readable.
> ```
> assertCommandFailure("add Valid Name p/12345 e/valid@email.butNoAddressPrefix valid, address",
>                                                                                expectedMessage);
> ```

Personally I don't like this. In the above example the second line (`expectedMessage);`) looks right-aligned, while the first line, as well as the majority of our code, will be left-aligned. This will lead to awkward breaks when reading the code since (english) readers will be accustomed to reading left-to-right rather than "left-and-look-for-the-end-of-the-sentence".

So, I'm wondering why you feel this may be more readable :-).

> However, it is a lot of extra work (for the dev to do as well as reviewer to explain :-p) and makes future automated refactoring messier (because the IDE doesn't honor/understand the extra indentation).

Both good points against this style :-P

> The alternative is to just accept the default 8 space indentation and live with slightly-harder-to-read code; easier to enforce and less work.

Yeah I would prefer this, and I don't feel that it is "harder-to-read".

damithc replied:
> I'm wondering why you feel this may be more readable 

Less eye travel, parameters nicely contained within brackets. :-)

![image](https://cloud.githubusercontent.com/assets/1673303/22024128/e0408b3a-dd03-11e6-94b3-039703bb3361.png)

But I agree, it may not be worth the trouble. I'll remove it from our coding standard.

damithc commented:
> So, I'm wondering why you feel this may be more readable

damithc commented:
Sorry, closed by mistake :-p

edmundmok commented:
@pyokagan PR updated, fixed the line breaks where possible.

edmundmok commented:
Should I update the title of the referenced issue, as well as this PR?

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -95,60 +96,84 @@ public void tearDown() {
 
         //Confirm the ui display elements should contain the right data
-        assertEquals(expectedMessage, result.feedbackToUser);
+        assertEquals(expectedMessage, feedbackToUser);
```
Does this work too (shorter by 5 lines)?

```java
        try {
           CommandResult result = logic.execute(inputCommand);
           asserEquals(expectedMessage, result.feedbackToUser);
        } catch (CommandException e) {
            assertTrue(expectCommandException);
            asserEquals(expectedMessage, e.getMessage());
        }
  ```

edmundmok replied:
@damithc It works if a `CommandException` is expected and indeed thrown. But if a `CommandException` is expected and not thrown, it would not enter that catch block and would not be detected.

damithc replied:
That's true. You can do the following? It's still shorter, but not sure if easy to understand. I'm ok with either way.

```java
        try {
           CommandResult result = logic.execute(inputCommand);
           assertFalse(expectCommandException);
           asserEquals(expectedMessage, result.feedbackToUser);
        } catch (CommandException e) {
            assertTrue(expectCommandException);
            asserEquals(expectedMessage, e.getMessage());
        }
  ```

BTW, `expectCommandException` --> `isCommandExcpetionExpected`?

edmundmok replied:
Yes, I should update `expectCommandException` to `isCommandExceptionExpected`, following the naming standards for booleans.

I'm ok with your suggested approach for the assertions. But I think it might be neater to keep all the assertions together and readable if both cases go through the same assertions...

Maybe @pyokagan has an idea about this?

pyokagan replied:
@damithc 
> expectCommandException --> isCommandExcpetionExpected?

Huh? Isn't that the naming convention for getters? e.g. if we had a variable named `commandExceptionExpected`, then its associated getter would be called `isCommandExceptionExpected`.

damithc replied:
Re naming convention for booleans, our current coding standard does not distinguish between methods, variables, (and parameters).  But I can see the point of phrasing a method as a question e.g. `isExceptionExpected()` and a variable/parameter as a fact `exceptionExpected` although the two are used similarly, making the differentiation mostly cosmetic. e.g. `if(exceptionExpected)` vs `if(isExceptionExpected())`.

The reason I prefer `is/has/...` prefix is that its more obvious. But I'm ok to adopt the two different styles too. In that case we need to update the coding standard.

pyokagan replied:
@damithc 
>That's true. You can do the following? It's still shorter, but not sure if easy to understand. I'm ok with either way.
>
>```java
>        try {
>           CommandResult result = logic.execute(inputCommand);
>           assertFalse(expectCommandException);
>           asserEquals(expectedMessage, result.feedbackToUser);
>        } catch (CommandException e) {
>            assertTrue(expectCommandException);
>            asserEquals(expectedMessage, e.getMessage());
>        }
> ```

Yeah this would certainly be more succinct. I'm fine with it.

Maybe we should also provide meaningful messages if the `expectCommandException` assertion fails, like:
```java
assertFalse("CommandException expected but was not thrown.", expectCommandException);
```

otherwise we won't have a meaningful error if the assertion fails.

pyokagan replied:
@damithc 
> our current coding standard does not distinguish between methods, variables, (and parameters).

I don't have a strong opinion yet, but I'd just like to clarify that the reason why I think only methods should have the `is` prefix and not field names is because it is consistent with the JavaBeans naming convention, which is supported by many Java libraries such as Jackson. So we may wish to adopt this convention as well to be able to work well with the rest of the Java ecosystem.

The JavaBeans convention doesn't say anything about parameter names though (?), but since parameters are closer to fields rather than methods, I guess they should follow the same naming conventions as fields.

damithc replied:
> Maybe we should also provide meaningful messages if the expectCommandException assertion fails

Yes, that makes the code more self-explanatory. 👍

damithc replied:
> I don't have a strong opinion yet, but I'd just like to clarify that the reason why I think only methods should have the is prefix and not field names is because it is consistent with the JavaBeans naming convention, which is supported by many Java libraries such as Jackson. So we may wish to adopt this convention as well to be able to work well with the rest of the Java ecosystem.

That's a good-enough reason. Let's update the coding convention after this semester is over (because it is used by CS2103 and already in-use for this week's tutorial). 

@edmundmok For consistency, use `is...` here. When the coding standard is updated, we'll have to apply it for the entire code base.

edmundmok replied:
Noted, will update it along with the assertions.

damithc replied:
@edmundmok I think you missed this comment from @pyokagan 

> Maybe we should also provide meaningful messages if the expectCommandException assertion fails, like: ...

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -95,60 +96,84 @@ public void tearDown() {
-     * Executes the command and confirms that the result message is correct and
-     * also confirms that the following three parts of the LogicManager object's state are as expected:<br>
+     * Executes the command, confirms that a CommandException is thrown and that the result message is correct.
+     * Both the 'address book' and the 'last shown list' are expected to be unchanged.
```
`expected to be` -> `verified to be`?

edmundmok replied:
Sounds good, will update this.

edmundmok commented:
PR updated.

edmundmok commented:
Oh yes. Sorry about that. Will update asap.

edmundmok commented:
Updated with messages for assertions on `isCommandExceptionExpected`

damithc commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -1,22 +1,53 @@
+
+    @Before
+    public void setUp() {
+        expectedStyleClass = new ArrayList<>(commandBox.getStyleClass());
```
`normalStyleOfCommandBox` is a better name? The term `expected` is usually used to indicate a temporary value that is expected at that point of the test while this value seems to be a fixed value.

`assertEquals(expectedStyleClass, commandBox.getStyleClass())` reads 'ensure the current style is the expected style'

`assertEquals(normalStyleOfCommandBox, commandBox.getStyleClass())` reads 'ensure the current style is the normal command box style`

The second one is more informative?

edmundmok replied:
```java
    private static final String ERROR_STYLE_CLASS = "error";
    private static final String COMMAND_THAT_SUCCEEDS = "select 3";
    private static final String COMMAND_THAT_FAILS = "invalid command";

    private ArrayList<String> defaultStyleOfCommandBox;
    private ArrayList<String> errorStyleOfCommandBox;

    @Before
    public void setUp() {
        defaultStyleOfCommandBox = new ArrayList<>(commandBox.getStyleClass());
        assertFalse(defaultStyleOfCommandBox.contains(ERROR_STYLE_CLASS));

        errorStyleOfCommandBox = new ArrayList<>(defaultStyleOfCommandBox);
        errorStyleOfCommandBox.add(ERROR_STYLE_CLASS);
        assertTrue(errorStyleOfCommandBox.contains(ERROR_STYLE_CLASS));
    }

    @Test
    public void commandBox_commandSucceeds_textClearedAndStyleClassRemainsTheSame() {
        commandBox.runCommand(COMMAND_THAT_SUCCEEDS);

        assertEquals("", commandBox.getCommandInput());
        assertEquals(defaultStyleOfCommandBox, commandBox.getStyleClass());
    }

    @Test
    public void commandBox_commandFails_textStaysAndErrorStyleClassAdded() {
        commandBox.runCommand(COMMAND_THAT_FAILS);

        assertEquals(COMMAND_THAT_FAILS, commandBox.getCommandInput());
        assertEquals(errorStyleOfCommandBox, commandBox.getStyleClass());
    }

    @Test
    public void commandBox_commandSucceedsAfterFailedCommand_textClearedAndErrorStyleClassRemoved() {
        // add error style to simulate a failed command
        commandBox.getStyleClass().add(ERROR_STYLE_CLASS);

        commandBox.runCommand(COMMAND_THAT_SUCCEEDS);

        assertEquals("", commandBox.getCommandInput());
        assertEquals(defaultStyleOfCommandBox, commandBox.getStyleClass());
    }
```
Agree that it seems to be more informative, will update.

damithc commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -1,22 +1,53 @@
+        commandBox.runCommand(COMMAND_THAT_FAILS);
+
+        assertEquals(COMMAND_THAT_FAILS, commandBox.getCommandInput());
+        expectedStyleClass.add(ERROR_STYLE_CLASS);
```
Define a separate `errorStyleofCommandBox`?

damithc commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -1,22 +1,53 @@
-        //TODO: confirm the text box color turns to red
+    public void commandBox_commandSucceedsAfterFailedCommand_textClearedAndErrorStyleClassRemoved() {
+        // add error style to simulate a failed command
+        commandBox.getStyleClass().add(ERROR_STYLE_CLASS);
```
Should this be a `commandBox.setErrorStyle()`  instead?

edmundmok replied:
Would probably be clearer

damithc commented on src/test/java/guitests/guihandles/CommandBoxHandle.java:
```diff
@@ -43,4 +44,12 @@ public HelpWindowHandle runHelpCommand() {
+        return getNode(COMMAND_INPUT_FIELD_ID).getStyleClass();
+    }
+
+    public void setErrorStyle() {
```
My bad, `setStyleToError` may be a better phrasing?

edmundmok replied:
Noted, will update.

damithc commented:
@pyokagan can have a look? Some minor changes after your last approval.

damithc commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -1,22 +1,58 @@
+        defaultStyleOfCommandBox = new ArrayList<>(commandBox.getStyleClass());
+        assertFalse(defaultStyleOfCommandBox.contains(ERROR_STYLE_CLASS));
+
+        errorStyleOfCommandBox = new ArrayList<>(defaultStyleOfCommandBox);
```
This line bothers me a bit because what it says is inaccurate (which is corrected in the next line). Temporary incorrect info affects readability.
May be squash the two lines like this:
`errorStyleOfCommandBox = (new ArrayList<>(defaultStyleOfCommandBox)).add(ERROR_STYLE_CLASS);`

edmundmok replied:
Hmm, that's true.

edmundmok replied:
`.add(...)` returns a boolean unfortunately. So we can't add and assign together at the same time.

damithc replied:
Yes, I forgot about that.
Use a temp variable? e.g. `copyOfDefaultStyle`

damithc replied:
One simple solution (just add a comment)
```java

// build style class for error
errorStyleOfCommandBox = new ArrayList<>(defaultStyleOfCommandBox);
errorStyleOfCommandBox.add(ERROR_STYLE_CLASS);
assertTrue(errorStyleOfCommandBox.contains(ERROR_STYLE_CLASS));

```

edmundmok replied:
Alright, I think that would suffice for now..

pyokagan commented on src/test/java/guitests/guihandles/CommandBoxHandle.java:
```diff
@@ -43,4 +44,12 @@ public HelpWindowHandle runHelpCommand() {
+    }
+
+    public void setStyleToError() {
+        getStyleClass().add("error");
```
Disagree with @damithc, I don't think this method would really be used in tests outside of `CommandBoxTest`.

Besides, this is repeating `ERROR_STYLE_CLASS` in `CommandBoxTest`. I'd really like to prevent the duplication of style class names as much as possible. (e.g. only one in `CommandBoxTest` and one in `CommandBox`)

damithc replied:
@pyokagan Agree about not repeating  `ERROR_STYLE_CLASS`. How about mentioning it only once in `CommandBox` and not in the test class? i.e. all queries about, and actions to, command box is done via `CommandBox` handle.  The test case uses a higher level language to simply ask CommandBox if the style is changed to error style but doesn't care what the error style is. Providing such a high-level API to query the SUT was the intention of the GuiHandle classes.

pyokagan replied:
@damithc 
> i.e. all queries about, and actions to, command box is done via CommandBoxHandle.

Yeah, I can see that it is the intention of the `*GuiHandle classes`. It makes me uneasy though, since we are technically not testing the SUT directly. It is like having an `AddressBookTest`, but then testing `ModelManager` instead, and concluding that `AddressBook` works because `ModelManager` does.

In other words, placing an abstraction layer between the test and the SUT provides a potential source of bugs.

damithc replied:
>  placing an abstraction layer between the test and the SUT provides a potential source of bugs.

Agreed, there is a risk of false positives (false negatives too, but they are less of a problem because they are detected immediately). This is a compromise to reduce complexity/duplication in test code. We had a similar problem in TEAMMATES and eventually came across the [Page Object](https://martinfowler.com/bliki/PageObject.html) pattern that helped to get a better handle of things. The GuiHandle solution is an adaptation of the same pattern. 

So, now that we have GuiHandle classes, it's better to go all the way and achieve a clear separation between the two? Even if we do, we should keep in mind the risk you pointed out and try to minimize logic in GuiHandle classes.

pyokagan replied:
@damithc 
> We had a similar problem in TEAMMATES and eventually came across the Page Object pattern that helped to get a better handle of things.

Thanks for the link, it helps put things in context.

> So, now that we have GuiHandle classes, it's better to go all the way and achieve a clear separation between the two?

Yes, let's go with that. It's better to have a clear application of a single pattern rather than a mixmash.

> Even if we do, we should keep in mind the risk you pointed out and try to minimize logic in GuiHandle classes.

Another thing we could do is to actually implement tests *for* GuiHandles, to ensure that the abstraction layer works as expected. These tests may still involve accessing some internal stuff directly in order to properly verify the correctness of the GuiHandles, but then we would have a rock-solid high-level abstraction layer which other tests can then build upon without needing to access these internals their own selves.

In the context of this PR, though, I do wonder how we would be able to implement our tests. In the current state of this PR, we strictly check that the CommandBox *only* appends the `error` style class whenever an error occurs (i.e preserves the order of other style classes), and *only* removes the `error` style class on a subsequent successful command. I hope that when we move all the query and action logic into `CommandBoxHandle* we would be able to preserve the strictness of these checks.

 but I'll leave @edmundmok to figure that out :-)

damithc replied:
> Another thing we could do is to actually implement tests for GuiHandles, to ensure that the abstraction layer works as expected. These tests may still involve accessing some internal stuff directly in order to properly verify the correctness of the GuiHandles, but then we would have a rock-solid high-level abstraction layer which other tests can then build upon without needing to access these internals their own selves.

Yup, let's plan to do that, in a separate PR of course.
@edmundmok can proceed with this PR..

pyokagan commented on src/test/java/guitests/CommandBoxTest.java:
```diff
@@ -1,22 +1,59 @@
+        // build style class for error
+        errorStyleOfCommandBox = new ArrayList<>(defaultStyleOfCommandBox);
+        errorStyleOfCommandBox.add(ERROR_STYLE_CLASS);
+        assertTrue(errorStyleOfCommandBox.contains(ERROR_STYLE_CLASS));
```
No need to include this, we can assume that the above `errorStyleOfCommandBox.add()` does what it is supposed to (since this test is not about testing `ObservableLists`)

edmundmok replied:
> (since this test is not about testing ObservableLists)

Good point. The first assert has a reason but this one doesn't really have one. Will remove this.

pyokagan replied:
@edmundmok 
> The first assert has a reason

Speaking of which, we should take advantage of the `message` argument of `assertFalse()` to explain that reason as well :-)

damithc commented:
Almost there @edmundmok Let's wrap this up. :-)

edmundmok commented:
@damithc @pyokagan Sorry for the inactivity, been quite busy. Will finish this up when I can find the time...

damithc commented:
Some other ongoing PRs can conflict with this one, so it is best to merge this before those PRs (and let them deal with the conflicts :-p). In the interest of merging quickly, perhaps any non-critical pending changes can be spawned off as `d.FirstTimer` issues?

pyokagan commented:
@damithc 
> Some other ongoing PRs can conflict with this one, so it is best to merge this before those PRs (and let them deal with the conflicts :-p).

Okay, maybe we could squash this fixup on top and merge this PR?

Changes:

* remove the `setStyleToError()` for now, otherwise we would have an awkward mix of "clear separation of GuiHandle responsibilities" and "not-very-clear separation of GuiHandle responsibilities" inside the `CommandBoxTest` code.

* Extracted `ERROR_STYLE_CLASS` into a public constant in `CommandBox`

* Add `assertFalse()` error message

```diff
diff --git a/src/main/java/seedu/address/ui/CommandBox.java b/src/main/java/seedu/address/ui/CommandBox.java
index bfdee4f9..70f7e9db 100644
--- a/src/main/java/seedu/address/ui/CommandBox.java
+++ b/src/main/java/seedu/address/ui/CommandBox.java
@@ -17,6 +17,7 @@ import java.util.logging.Logger;
 public class CommandBox extends UiPart<Region> {
     private final Logger logger = LogsCenter.getLogger(CommandBox.class);
     private static final String FXML = "CommandBox.fxml";
+    public static final String ERROR_STYLE_CLASS = "error";
 
     private final Logic logic;
 
@@ -60,14 +61,14 @@ public class CommandBox extends UiPart<Region> {
      * Sets the command box style to indicate a successful command.
      */
     private void setStyleToIndicateCommandSuccess() {
-        commandTextField.getStyleClass().remove("error");
+        commandTextField.getStyleClass().remove(ERROR_STYLE_CLASS);
     }
 
     /**
      * Sets the command box style to indicate a failed command.
      */
     private void setStyleToIndicateCommandFailure() {
-        commandTextField.getStyleClass().add("error");
+        commandTextField.getStyleClass().add(ERROR_STYLE_CLASS);
     }
 
 }
diff --git a/src/test/java/guitests/CommandBoxTest.java b/src/test/java/guitests/CommandBoxTest.java
index b302efef..71efa5d1 100644
--- a/src/test/java/guitests/CommandBoxTest.java
+++ b/src/test/java/guitests/CommandBoxTest.java
@@ -3,15 +3,14 @@ package guitests;
 import org.junit.Before;
 import org.junit.Test;
 
+import seedu.address.ui.CommandBox;
+
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertTrue;
-
 import java.util.ArrayList;
 
 public class CommandBoxTest extends AddressBookGuiTest {
 
-    private static final String ERROR_STYLE_CLASS = "error";
     private static final String COMMAND_THAT_SUCCEEDS = "select 3";
     private static final String COMMAND_THAT_FAILS = "invalid command";
 
@@ -21,12 +20,12 @@ public class CommandBoxTest extends AddressBookGuiTest {
     @Before
     public void setUp() {
         defaultStyleOfCommandBox = new ArrayList<>(commandBox.getStyleClass());
-        assertFalse(defaultStyleOfCommandBox.contains(ERROR_STYLE_CLASS));
+        assertFalse("CommandBox default style classes should not contain error style class.",
+                    defaultStyleOfCommandBox.contains(CommandBox.ERROR_STYLE_CLASS));
 
         // build style class for error
         errorStyleOfCommandBox = new ArrayList<>(defaultStyleOfCommandBox);
-        errorStyleOfCommandBox.add(ERROR_STYLE_CLASS);
-        assertTrue(errorStyleOfCommandBox.contains(ERROR_STYLE_CLASS));
+        errorStyleOfCommandBox.add(CommandBox.ERROR_STYLE_CLASS);
     }
 
     @Test
@@ -48,7 +47,7 @@ public class CommandBoxTest extends AddressBookGuiTest {
     @Test
     public void commandBox_commandSucceedsAfterFailedCommand_textClearedAndErrorStyleClassRemoved() {
         // add error style to simulate a failed command
-        commandBox.setStyleToError();
+        commandBox.getStyleClass().add(CommandBox.ERROR_STYLE_CLASS);
 
         commandBox.runCommand(COMMAND_THAT_SUCCEEDS);
 
diff --git a/src/test/java/guitests/guihandles/CommandBoxHandle.java b/src/test/java/guitests/guihandles/CommandBoxHandle.java
index 287c5c9e..bc57cbb6 100644
--- a/src/test/java/guitests/guihandles/CommandBoxHandle.java
+++ b/src/test/java/guitests/guihandles/CommandBoxHandle.java
@@ -49,7 +49,4 @@ public class CommandBoxHandle extends GuiHandle {
         return getNode(COMMAND_INPUT_FIELD_ID).getStyleClass();
     }
 
-    public void setStyleToError() {
-        getStyleClass().add("error");
-    }
 }
```

damithc commented:
> maybe we could squash this fixup on top and merge this PR?

Sure, let's do that. 👍

edmundmok commented:
Updated according to @pyokagan 's changes.

damithc commented:
@edmundmok branch gone out of sync. Can do a rebase?

pyokagan commented:
@damithc 
> @edmundmok branch gone out of sync. Can do a rebase?

Wouldn't it just be sufficient to manually re-run Travis/AppVeyor? They should use GitHub's updated merge commits, I think.

damithc commented:
> Wouldn't it just be sufficient to manually re-run Travis/AppVeyor? They should use GitHub's updated merge commits, I think.

Let me try that.

damithc commented:
Good job @edmundmok 👍 This is a notable improvement to the design as well as the external behavior. 
For future reference, the PR title should be same as the issue it fixes, **including the issue number**. I've updated the title of this PR.

As usual, kudos to @pyokagan for all the review input over the course of this PR.
